libraries/bot-builder/src/main/java/com/microsoft/bot/builder/CloudAdapterBase.java

```diff
@@ -0,0 +1,456 @@
+     * @return The {@link BotFrameworkAuthentication} instance for this adapter.
+     */
+    public BotFrameworkAuthentication getBotFrameworkAuthentication() {
+        return botFrameworkAuthentication;
+    }
+
+    /**
+     * Initializes a new instance of the {@link CloudAdapterBase} class.
+     * @param withBotFrameworkAuthentication The cloud environment used for validating and creating tokens.
+     */
+    protected CloudAdapterBase(BotFrameworkAuthentication withBotFrameworkAuthentication) {
+        if (withBotFrameworkAuthentication == null) {
+            throw new IllegalArgumentException("withBotFrameworkAuthentication cannot be null");
+        }
+        this.botFrameworkAuthentication = withBotFrameworkAuthentication;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<ResourceResponse[]> sendActivities(TurnContext context, List<Activity> activities) {
+        if (context == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("context"));
+        }
+
+        if (activities == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("activities"));
+        }
+
+        if (activities.size() == 0) {
+            return Async.completeExceptionally(
+                new IllegalArgumentException("Expecting one or more activities, but the array was empty.")
+            );
+        }
+
+        logger.info(String.format("sendActivities for %d activities.", activities.size()));
+
+        return CompletableFuture.supplyAsync(() -> {
+            ResourceResponse[] responses = new ResourceResponse[activities.size()];
+
+            for (int index = 0; index < activities.size(); index++) {
+                Activity activity = activities.get(index);
+
+                activity.setId(null);
+                ResourceResponse response;
+
+                logger.info(String.format("Sending activity. ReplyToId: %s", activity.getReplyToId()));
+
+                if (activity.isType(ActivityTypes.DELAY)) {
+                    int delayMs = (int) activity.getValue();
+                    try {
+                        Thread.sleep(delayMs);
+                    } catch (InterruptedException e) {
+                        throw new RuntimeException(e);
+                    }
+                    response = null;
+                } else if (activity.isType(ActivityTypes.INVOKE_RESPONSE)) {
+                    context.getTurnState().add(BotFrameworkAdapter.INVOKE_RESPONSE_KEY, activity);
+                    response = null;
+                } else if (
+                    activity.isType(ActivityTypes.TRACE)
+                        && !StringUtils.equals(activity.getChannelId(), Channels.EMULATOR)
+                ) {
+                    // no-op
+                    response = null;
+                } else if (StringUtils.isNotBlank(activity.getReplyToId())) {
```

Comment:
I see you simplified an internal if and else here with an else if and then another else. I guess you did it for simplification. I understand it's ok 🤔

Reply:
Yes, also, I took the same implementation from [BotFrameworkAdapter](https://github.com/microsoft/botbuilder-java/blob/main/libraries/bot-builder/src/main/java/com/microsoft/bot/builder/BotFrameworkAdapter.java#L564) as it was the same code